File: dat-pypi/tests/test_error_code.py
import time

import pytest

from saro_dat import DatCertificate, DatCrypto, DatError, DatManager, DatRetry, DatSignature
from saro_dat import error as E
from saro_dat.crypto import DatCryptoAlgorithm
from saro_dat.signature import DatSignatureAlgorithm

SIG = DatSignatureAlgorithm.ECDSA_P256
CRY = DatCryptoAlgorithm.IV_AES256_GCM


def certificate(cid=1, start_offset=-10, duration=200, ttl=100):
    now = int(time.time())
    return DatCertificate(
        cid, now + start_offset, duration, ttl,
        DatSignature.generate(SIG), DatCrypto.generate(CRY),
    )


def issuable_manager(cid=1):
    manager = DatManager()
    manager.import_certificates([certificate(cid)], True)
    return manager


def code_of(fn):
    with pytest.raises(DatError) as excinfo:
        fn()
    return excinfo.value.code


def test_expired_token_is_not_malformed():
    manager = issuable_manager()
    token = manager.issue("p", "s")
    rest = token.split(".", 1)[1]
    now = int(time.time())

    assert code_of(lambda: manager.parse(f"{now - 1}.{rest}")) == E.TOKEN_EXPIRED
    assert code_of(lambda: manager.parse(f"{now}.{rest}")) == E.TOKEN_EXPIRED


def test_malformed_token_shapes():
    manager = issuable_manager()
    token = manager.issue("p", "s")
    parts = token.split(".")

    assert code_of(lambda: manager.parse("1.2.3")) == E.TOKEN_MALFORMED
    assert code_of(lambda: manager.parse(token + ".extra")) == E.TOKEN_MALFORMED
    assert code_of(lambda: manager.parse("+" + token)) == E.TOKEN_MALFORMED
    assert code_of(lambda: manager.parse(".".join([parts[0], "zz", *parts[2:]]))) == E.TOKEN_MALFORMED


def test_empty_signature_is_sig_malformed():
    manager = issuable_manager()
    token = manager.issue("p", "s")
    parts = token.split(".")

    assert code_of(lambda: manager.parse(".".join(parts[:4]) + ".")) == E.SIG_MALFORMED


def test_forged_signature_is_sig_mismatch():
    victim = issuable_manager(7)
    attacker = issuable_manager(7)
    forged = attacker.issue("p", "s")

    with pytest.raises(DatError) as excinfo:
        victim.parse(forged)
    assert excinfo.value.code == E.SIG_MISMATCH
    assert excinfo.value.security_event is True
    assert excinfo.value.retry is DatRetry.PERMANENT


def test_tampered_secure_is_crypto_tag_mismatch():
    manager = issuable_manager()
    cert = manager._certificates_by_cid[1]
    token = DatManager._issue(cert, "plain", "secure-payload")
    parts = token.split(".")

    secure = parts[3]
    parts[3] = secure[:-1] + ("B" if secure[-1] == "A" else "A")

    from saro_dat import Dat
    dat = Dat(".".join(parts))
    with pytest.raises(DatError) as excinfo:
        cert._crypto_key.decrypt(dat._secure)
    assert excinfo.value.code == E.CRYPTO_TAG_MISMATCH
    assert excinfo.value.security_event is True


def test_unknown_cid_is_cert_not_found():
    manager = issuable_manager(1)
    other = issuable_manager(999)
    token = other.issue("p", "s")

    assert code_of(lambda: manager.parse(token)) == E.CERT_NOT_FOUND


def test_duplicate_cid_on_import():
    manager = DatManager()
    certs = [certificate(5), certificate(5)]
    assert code_of(lambda: manager.import_certificates(certs, True)) == E.CERT_DUPLICATE_CID


def test_malformed_certificate_shapes():
    assert code_of(lambda: DatCertificate.imports("a.b.c")) == E.CERT_MALFORMED
    assert code_of(
        lambda: DatCertificate.imports("zz.1.2.3.ECDSA-P256.IV-AES256-GCM.AAAA.AAAA")
    ) == E.CERT_MALFORMED


def test_no_certificate_at_all():
    with pytest.raises(DatError) as excinfo:
        DatManager().issue("p", "s")
    assert excinfo.value.code == E.MANAGER_NO_CERTIFICATE
    assert excinfo.value.retry is DatRetry.TRANSIENT


def test_issuance_window_not_yet_open_is_transient():
    manager = DatManager()
    manager.import_certificates([certificate(1, start_offset=3600)], True)

    with pytest.raises(DatError) as excinfo:
        manager.issue("p", "s")
    assert excinfo.value.code == E.MANAGER_NO_ISSUABLE_CERTIFICATE
    assert excinfo.value.__cause__.code == E.CERT_NOT_YET_ISSUABLE
    assert excinfo.value.retry is DatRetry.TRANSIENT


def test_issuance_window_closed_is_permanent():
    manager = DatManager()
    manager.import_certificates([certificate(1, start_offset=-500, duration=100, ttl=3600)], True)

    with pytest.raises(DatError) as excinfo:
        manager.issue("p", "s")
    assert excinfo.value.code == E.MANAGER_NO_ISSUABLE_CERTIFICATE
    assert excinfo.value.__cause__.code == E.CERT_ISSUANCE_ENDED
    assert excinfo.value.retry is DatRetry.PERMANENT


def test_verify_only_certificate_cannot_issue():
    source = certificate(1)
    verify_only = DatCertificate.imports(source.exports(True))
    manager = DatManager()
    manager.import_certificates([verify_only], True)

    with pytest.raises(DatError) as excinfo:
        manager.issue("p", "s")
    assert excinfo.value.code == E.MANAGER_NO_ISSUABLE_CERTIFICATE
    assert excinfo.value.__cause__.code == E.CERT_VERIFY_ONLY
    assert excinfo.value.retry is DatRetry.PERMANENT


def test_unknown_algorithm_names():
    assert code_of(lambda: DatSignature.generate("NOPE")) == E.CONFIG_ALG_UNSUPPORTED
    assert code_of(lambda: DatCrypto.generate("NOPE")) == E.CONFIG_ALG_UNSUPPORTED


def test_wrong_key_size_is_key_invalid():
    assert code_of(lambda: DatCrypto.imports(CRY.value, "AAAA")) == E.KEY_INVALID
    assert code_of(lambda: DatSignature.imports("HMAC-SHA256-MFS", "AAAA")) == E.KEY_INVALID
    assert code_of(lambda: DatSignature.imports(SIG, "AAAA")) == E.KEY_INVALID


def test_hmac_verify_only_export_is_structurally_unsupported():
    hmac = DatSignature.generate(DatSignatureAlgorithm.HMAC_SHA256_MFS)
    assert code_of(lambda: hmac.exports(True)) == E.KEY_VERIFY_ONLY_UNSUPPORTED


def test_signing_with_verify_only_key_is_key_missing():
    source = DatSignature.generate(SIG)
    public_only = DatSignature.imports(SIG, source.exports(True))
    assert code_of(lambda: public_only.sign("body")) == E.SIG_KEY_MISSING


def test_ciphertext_shorter_than_iv():
    crypto = DatCrypto.generate(CRY)
    assert code_of(lambda: crypto.decrypt(b"12345")) == E.CRYPTO_DATA_INVALID


def test_empty_secure_payload_is_not_an_error():
    crypto = DatCrypto.generate(CRY)
    assert crypto.encrypt(b"") == b""
    assert crypto.decrypt(b"") == b""


def test_codes_are_well_formed():
    samples = [
        E.TOKEN_MALFORMED, E.TOKEN_EXPIRED, E.CERT_EXPIRED, E.CERT_NOT_SYNCED,
        E.SIG_MISMATCH, E.CRYPTO_TAG_MISMATCH, E.KEY_INVALID,
        E.MANAGER_NO_CERTIFICATE, E.CMS_UNAUTHORIZED, E.CMS_SYNC_IN_PROGRESS,
        E.CONFIG_ALG_UNSUPPORTED, E.INTERNAL_UNAVAILABLE,
    ]
    for code in samples:
        assert code.startswith("DAT_"), code
        assert all(c.isupper() or c == "_" for c in code), code
        assert str(DatError(code)).startswith(code)
        assert str(DatError(code, "detail")) == f"{code}: detail"


def test_dat_error_is_catchable_as_both_legacy_types():
    e = DatError(E.TOKEN_MALFORMED)
    assert isinstance(e, ValueError)
    assert isinstance(e, RuntimeError)


def test_retry_classification():
    for code in (E.CMS_UNAUTHORIZED, E.CMS_FORBIDDEN, E.CMS_ENDPOINT_NOT_FOUND):
        assert DatError(code).retry is DatRetry.PERMANENT, code
    for code in (E.CMS_UNREACHABLE, E.CMS_SERVER_ERROR, E.CMS_NOT_SYNCED):
        assert DatError(code).retry is DatRetry.TRANSIENT, code
    for code in (E.CMS_SYNC_IN_PROGRESS, E.CMS_VERSION_RESET):
        assert DatError(code).retry is DatRetry.STATE, code

File: dat-ruby/test/test_error_code.rb
# frozen_string_literal: true

require_relative 'test_helper'

class TestErrorCode < Minitest::Test
  EC = Saro::Dat::ErrorCode
  SIG = Saro::Dat::DatSignatureAlgorithm::ECDSA_P256
  CRY = Saro::Dat::DatCryptoAlgorithm::IV_AES256_GCM

  def certificate(cid, start_offset, duration, ttl)
    Saro::Dat::DatCertificate.generate(cid, Time.now.to_i + start_offset, duration, ttl, SIG, CRY)
  end

  def issuable_manager(cid = 1)
    m = Saro::Dat::DatManager.new
    m.import_certificates([certificate(cid, -10, 200, 100)])
    m
  end

  def code_of
    yield
    flunk "expected an error, got success"
  rescue Saro::Dat::Error => e
    e.code
  end

  def error_of
    yield
    flunk "expected an error, got success"
  rescue Saro::Dat::Error => e
    e
  end

  def with_expire(token, expire)
    "#{expire}.#{token.split('.', 2)[1]}"
  end

  def test_expired_token_is_not_malformed
    m = issuable_manager
    token = m.issue('p', 's')
    now = Time.now.to_i

    assert_equal EC::TOKEN_EXPIRED, code_of { m.parse(with_expire(token, now - 1)) }
    assert_equal EC::TOKEN_EXPIRED, code_of { m.parse(with_expire(token, now)) }
  end

  def test_malformed_token_shapes
    m = issuable_manager
    token = m.issue('p', 's')
    parts = token.split('.', -1)

    assert_equal EC::TOKEN_MALFORMED, code_of { m.parse('1.2.3') }
    assert_equal EC::TOKEN_MALFORMED, code_of { m.parse("#{token}.extra") }
    assert_equal EC::TOKEN_MALFORMED, code_of { m.parse("#{token}.") }
    assert_equal EC::TOKEN_MALFORMED, code_of { m.parse("+#{token}") }
    assert_equal EC::TOKEN_MALFORMED, code_of { m.parse([parts[0], 'zz', *parts[2..]].join('.')) }
  end

  def test_empty_signature_is_sig_malformed
    m = issuable_manager
    parts = m.issue('p', 's').split('.', -1)

    assert_equal EC::SIG_MALFORMED, code_of { m.parse("#{parts[0, 4].join('.')}.") }
  end

  def test_forged_signature_is_sig_mismatch
    victim = issuable_manager(7)
    attacker = issuable_manager(7)
    forged = attacker.issue('p', 's')

    e = error_of { victim.parse(forged) }
    assert_equal EC::SIG_MISMATCH, e.code
    assert e.security_event?, '위조는 보안 이벤트로 표시되어야 한다'
    assert_equal :permanent, e.retry
  end

  def test_tampered_secure_is_crypto_tag_mismatch
    cert = certificate(1, -10, 200, 100)
    token = Saro::Dat::DatManager._issue(cert, 'plain', 'secure-payload')
    parts = token.split('.', -1)

    last = parts[3][-1]
    parts[3] = parts[3][0..-2] + (last == 'A' ? 'B' : 'A')

    dat = Saro::Dat::Dat.new(parts.join('.'))
    e = error_of { cert.crypto_key.decrypt(dat.secure) }
    assert_equal EC::CRYPTO_TAG_MISMATCH, e.code
    assert e.security_event?
  end

  def test_unknown_cid_is_cert_not_found
    m = issuable_manager(1)
    other = issuable_manager(999)

    assert_equal EC::CERT_NOT_FOUND, code_of { m.parse(other.issue('p', 's')) }
  end

  def test_duplicate_cid_on_import
    a = certificate(5, -10, 200, 100)
    b = certificate(5, -10, 200, 100)

    assert_equal EC::CERT_DUPLICATE_CID, code_of { Saro::Dat::DatManager.new.import_certificates([a, b]) }
  end

  def test_malformed_certificate_shapes
    assert_equal EC::CERT_MALFORMED, code_of { Saro::Dat::DatCertificate.imports('a.b.c') }
    assert_equal EC::CERT_MALFORMED,
                 code_of { Saro::Dat::DatCertificate.imports('zz.1.2.3.ECDSA-P256.IV-AES256-GCM.AAAA.AAAA') }
    assert_equal EC::CERT_MALFORMED,
                 code_of { certificate_with_times(1, 0xFFFFFFFFFFFFFFFF, 1, 0) }
  end

  def certificate_with_times(cid, start, duration, ttl)
    Saro::Dat::DatCertificate.new(
      cid, start, duration, ttl,
      Saro::Dat::DatSignature.generate(SIG),
      Saro::Dat::DatCrypto.generate(CRY)
    )
  end

  def test_no_certificate_at_all
    e = error_of { Saro::Dat::DatManager.new.issue('p', 's') }
    assert_equal EC::MANAGER_NO_CERTIFICATE, e.code
    assert_equal :transient, e.retry
  end

  def test_issuance_window_not_yet_open_is_transient
    m = Saro::Dat::DatManager.new
    m.import_certificates([certificate(1, 3600, 200, 100)])

    e = error_of { m.issue('p', 's') }
    assert_equal EC::MANAGER_NO_ISSUABLE_CERTIFICATE, e.code
    assert_equal EC::CERT_NOT_YET_ISSUABLE, e.cause.code
    assert_equal :transient, e.retry
  end

  def test_issuance_window_closed_is_permanent
    m = Saro::Dat::DatManager.new
    m.import_certificates([certificate(1, -500, 100, 3600)])

    e = error_of { m.issue('p', 's') }
    assert_equal EC::MANAGER_NO_ISSUABLE_CERTIFICATE, e.code
    assert_equal EC::CERT_ISSUANCE_ENDED, e.cause.code
    assert_equal :permanent, e.retry
  end

  def test_verify_only_certificate_cannot_issue
    source = certificate(1, -10, 200, 100)
    verify_only = Saro::Dat::DatCertificate.imports(source.exports(true))
    m = Saro::Dat::DatManager.new
    m.import_certificates([verify_only])

    e = error_of { m.issue('p', 's') }
    assert_equal EC::MANAGER_NO_ISSUABLE_CERTIFICATE, e.code
    assert_equal EC::CERT_VERIFY_ONLY, e.cause.code
    assert_equal :permanent, e.retry
  end

  def test_unknown_algorithm_names
    assert_equal EC::CONFIG_ALG_UNSUPPORTED, code_of { Saro::Dat::DatSignature.generate('NOPE') }
    assert_equal EC::CONFIG_ALG_UNSUPPORTED, code_of { Saro::Dat::DatCrypto.generate('NOPE') }
  end

  def test_wrong_key_size_is_key_invalid
    assert_equal EC::KEY_INVALID, code_of { Saro::Dat::DatCrypto.imports(CRY, 'AAAA') }
    assert_equal EC::KEY_INVALID,
                 code_of { Saro::Dat::DatSignature.imports(Saro::Dat::DatSignatureAlgorithm::HMAC_SHA256_MFS, 'AAAA') }
    assert_equal EC::KEY_INVALID, code_of { Saro::Dat::DatSignature.imports(SIG, 'AAAA') }
  end

  def test_hmac_verify_only_export_is_structurally_unsupported
    hmac = Saro::Dat::DatSignature.generate(Saro::Dat::DatSignatureAlgorithm::HMAC_SHA256_MFS)
    assert_equal EC::KEY_VERIFY_ONLY_UNSUPPORTED, code_of { hmac.exports(true) }
  end

  def test_signing_with_verify_only_key_is_key_missing
    source = Saro::Dat::DatSignature.generate(SIG)
    public_only = Saro::Dat::DatSignature.imports(SIG, source.exports(true))

    assert_equal EC::SIG_KEY_MISSING, code_of { public_only.sign('body') }
  end

  def test_ciphertext_shorter_than_iv_and_tag
    crypto = Saro::Dat::DatCrypto.generate(CRY)
    assert_equal EC::CRYPTO_DATA_INVALID, code_of { crypto.decrypt('12345'.b) }
  end

  def test_empty_secure_payload_is_not_an_error
    crypto = Saro::Dat::DatCrypto.generate(CRY)
    assert_equal '', crypto.encrypt('')
    assert_equal '', crypto.decrypt('')
  end

  def test_every_code_is_well_formed
    codes = EC.constants.map { |c| EC.const_get(c) }.grep(String)
    refute_empty codes
    codes.each do |code|
      assert_match(/\A DAT_[A-Z_]+ \z/x, code, "#{code} must be DAT_ SCREAMING_SNAKE_CASE")
    end
  end

  def test_code_comes_first_in_the_message
    assert_equal 'DAT_TOKEN_EXPIRED', Saro::Dat::Error.new(EC::TOKEN_EXPIRED).message
    assert_equal 'DAT_TOKEN_MALFORMED: bad field',
                 Saro::Dat::Error.new(EC::TOKEN_MALFORMED, 'bad field').message
  end

  def test_retry_classification
    [EC::CMS_UNAUTHORIZED, EC::CMS_FORBIDDEN, EC::CMS_ENDPOINT_NOT_FOUND].each do |code|
      assert_equal :permanent, Saro::Dat::Error.new(code).retry, code
    end
    [EC::CMS_UNREACHABLE, EC::CMS_SERVER_ERROR, EC::CMS_NOT_SYNCED].each do |code|
      assert_equal :transient, Saro::Dat::Error.new(code).retry, code
    end
    [EC::CMS_SYNC_IN_PROGRESS, EC::CMS_VERSION_RESET].each do |code|
      assert_equal :state, Saro::Dat::Error.new(code).retry, code
    end
  end

  def test_http_status_maps_to_distinct_codes
    m = Saro::Dat::DatCmsManager
    assert_equal EC::CMS_UNAUTHORIZED, m.http_status_error(401).code
    assert_equal EC::CMS_FORBIDDEN, m.http_status_error(403).code
    assert_equal EC::CMS_ENDPOINT_NOT_FOUND, m.http_status_error(404).code
    assert_equal EC::CMS_SERVER_ERROR, m.http_status_error(503).code
    assert_equal EC::CMS_HTTP_STATUS, m.http_status_error(418).code
  end

  def test_cause_chain_is_preserved
    inner = Saro::Dat::Error.new(EC::CERT_MALFORMED, 'bad field')
    outer = Saro::Dat::Error.new(EC::CMS_IMPORT_FAILED, cause: inner)

    assert_equal EC::CMS_IMPORT_FAILED, outer.code
    assert_equal EC::CERT_MALFORMED, outer.cause.code
  end

  def test_uri_validation_uses_config_uri_invalid
    b = Saro::Dat::DatCmsManager.builder
    assert_equal EC::CONFIG_URI_INVALID, code_of { b.uri('http://localhost:8088/abc').build }
    assert_equal EC::CONFIG_URI_INVALID, code_of { Saro::Dat::DatCmsManager.builder.uri('ftp://localhost').build }
  end
end
